python/taichi/lang/matrix.py

```diff
@@ -85,11 +87,45 @@ def __init__(self,
                     if keep_raw:
                         mat = [list([x]) for x in n]
                     else:
-                        mat = [list([expr.Expr(x)]) for x in n]
+                        if not ti.is_extension_supported(
+                                ti.cfg.arch, ti.extension.dynamic_index
+                        ) or in_python_scope() or disable_local_tensor:
+                            mat = [list([expr.Expr(x)]) for x in n]
+                        else:
+                            dt = ti.default_cfg().default_ip if isinstance(
```

Comment:
I believe it's possible to have `x = ti.Vector([a, b, c])`, where `a, b, c` themselves are expressions. In that case we need to use their types

Reply:
https://github.com/taichi-dev/taichi/pull/2637/commits/607d2f11532e1bac42a6f01d2ec1041c70ecff1a